File: core/src/main/java/com/nagopy/android/overlayviewmanager/OverlayViewManager.java
/*
 * Copyright 2017 75py
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nagopy.android.overlayviewmanager;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.fragment.app.FragmentManager;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.WindowManager;

import com.nagopy.android.overlayviewmanager.internal.Logger;
import com.nagopy.android.overlayviewmanager.internal.OverlayWindowManager;
import com.nagopy.android.overlayviewmanager.internal.ScreenMonitor;

/**
 * Overlay your views!
 */
public final class OverlayViewManager {

    Context applicationContext;
    WindowManager windowManager;

    static OverlayViewManager INSTANCE = new OverlayViewManager();

    /**
     * Initialize.
     *
     * @param application Your {@link Application} instance
     */
    public static void init(Application application) {
        INSTANCE.initialize(application);
    }

    /**
     * Return the {@link OverlayViewManager} instance. The instance is singleton.
     *
     * @return instance
     */
    public static OverlayViewManager getInstance() {
        if (INSTANCE.applicationContext == null) {
            throw new IllegalStateException("OverlayViewManager is not initialized. Please call initApplicationInstance(Application).");
        }
        return INSTANCE;
    }

    OverlayViewManager() {
    }

    /**
     * Initialize. Please call this method before {@link OverlayView} use.
     *
     * @param context {@link Context}
     */
    void initialize(@NonNull Context context) {
        applicationContext = context.getApplicationContext();
        windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        OverlayWindowManager.initApplicationInstance(windowManager);
        ScreenMonitor.init(applicationContext);
    }

    /**
     * Wrapper of {@link Settings#canDrawOverlays(Context)}.
     *
     * @return true: This app context is allowed display over other apps
     */
    public boolean canDrawOverlays() {
        //noinspection SimplifiableIfStatement
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        } else {
            return Settings.canDrawOverlays(applicationContext);
        }
    }

    /**
     * Show permission request dialog.
     *
     * @param fragmentManager {@link FragmentManager}
     * @param appNameId       String resource id of application name
     */
    @TargetApi(Build.VERSION_CODES.M)
    public void showPermissionRequestDialog(@NonNull FragmentManager fragmentManager, @StringRes int appNameId) {
        PermissionRequestDialogFragment dialog = PermissionRequestDialogFragment.newInstance(appNameId);
        dialog.show(fragmentManager, "PermissionRequestDialogFragment");
    }

    /**
     * Request overlay permission. Open the setting app.
     */
    @TargetApi(Build.VERSION_CODES.M)
    public void requestOverlayPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (!canDrawOverlays()) {
                Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION
                        , Uri.parse("package:" + applicationContext.getPackageName()));
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                applicationContext.startActivity(intent);
            }
        } else {
            Logger.d("API level %d is not required any user action.", Build.VERSION.SDK_INT);
        }
    }

    /**
     * Get display width.
     *
     * @return Display width(px)
     */
    public int getDisplayWidth() {
        DisplayMetrics metrics = new DisplayMetrics();
        windowManager.getDefaultDisplay().getMetrics(metrics);
        return metrics.widthPixels;
    }

    /**
     * Get display height. It includes the status bar and the navigation bar.
     *
     * @return Display height(px)
     */
    public int getDisplayHeight() {
        DisplayMetrics metrics = new DisplayMetrics();
        windowManager.getDefaultDisplay().getMetrics(metrics);
        return metrics.heightPixels;
    }

    /**
     * Create a new OverlayView instance. ViewScope = ACTIVITY
     * If you want to show your view in Activity's lifecycle, use #newOverlayView(View, Activity).
     * Otherwise(If you want to show your view regardless of whether Activity is alive), use #newOverlayView(View).
     * <p>
     * If you use OverlayView from Activity only, you don't have to get "Draw over other apps" permission.
     *
     * @param view     The view that you want to overlay
     * @param <T>      View
     * @param activity Parent activity
     * @return A new instance of OverlayView
     */
    @NonNull
    public <T extends View> OverlayView<T> newOverlayView(T view, Activity activity) {
        return OverlayView.create(view, activity);
    }

    /**
     * Create a new OverlayView instance. ViewScope = APPLICATION
     * If you want to show your view regardless of whether Activity is alive, use #newOverlayView(View).
     * Otherwise(if you want to show your view in Activity's lifecycle), use #newOverlayView(View, Activity).
     *
     * @param view The view that you want to overlay
     * @param <T>  View
     * @return A new instance of OverlayView
     */
    @NonNull
    public <T extends View> OverlayView<T> newOverlayView(T view) {
        return OverlayView.create(view);
    }
}

File: core/src/main/kotlin/com/nagopy/android/overlayviewmanager/OverlayViewManager.kt
/*
 * Copyright 2026 75py
 * Licensed under the Apache License, Version 2.0 (the "License");
 */
package com.nagopy.android.overlayviewmanager

import android.app.Activity
import android.app.Application
import android.annotation.TargetApi
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Looper
import android.view.View
import android.view.WindowManager
import androidx.annotation.MainThread
import androidx.annotation.StringRes
import androidx.fragment.app.FragmentManager
import com.nagopy.android.overlayviewmanager.internal.ActivityOverlayRegistry
import com.nagopy.android.overlayviewmanager.internal.OverlayWindowManager
import com.nagopy.android.overlayviewmanager.internal.SimpleActivityLifecycleCallbacks

/** Process-scoped factory for synchronous overlay handles. */
public class OverlayViewManager private constructor(private val application: Application) {
    private val windowManager: WindowManager = application.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    private val permission: OverlayPermission = OverlayPermission()

    init {
        OverlayWindowManager.initApplicationInstance(windowManager)
    }

    /** Creates an application-scoped overlay with default immutable configuration. */
    @MainThread public fun <T : View> newOverlayView(view: T): OverlayView<T> = newApplicationOverlay(view, OverlaySpec())
    /** Creates an application-scoped overlay with [spec]. */
    @MainThread public fun <T : View> newOverlayView(view: T, spec: OverlaySpec): OverlayView<T> = newApplicationOverlay(view, spec)
    /** Creates an activity-scoped overlay with default immutable configuration. */
    @MainThread public fun <T : View> newOverlayView(view: T, activity: Activity): OverlayView<T> = newActivityOverlay(view, activity, OverlaySpec())
    /** Creates an activity-scoped overlay with [spec]. */
    @MainThread public fun <T : View> newOverlayView(view: T, activity: Activity, spec: OverlaySpec): OverlayView<T> = newActivityOverlay(view, activity, spec)

    /** Returns the immutable overlay permission helper. See [OverlayPermission]. */
    public fun overlayPermission(): OverlayPermission = permission

    /** Temporary 2.x bridge. New code should use [OverlayPermission.isGranted] instead. */
    @Deprecated("Use OverlayPermission.isGranted from the host application instead.") public fun canDrawOverlays(): Boolean = overlayPermission().isGranted(application)

    /**
     * Temporary 2.x bridge retained until sample and consumer migration.
     * New code should use [OverlayPermission.settingsIntent] and own its rationale UI instead.
     */
    @Deprecated("Use OverlayPermission.settingsIntent from the host application instead.")
    @TargetApi(Build.VERSION_CODES.M)
    public fun showPermissionRequestDialog(fragmentManager: FragmentManager, @StringRes appNameId: Int) {
        PermissionRequestDialogFragment.newInstance(appNameId).show(fragmentManager, "PermissionRequestDialogFragment")
    }

    /** Temporary 2.x bridge retained until sample and consumer migration. */
    @Deprecated("Use OverlayPermission.settingsIntent from the host application instead.")
    @TargetApi(Build.VERSION_CODES.M)
    public fun requestOverlayPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !canDrawOverlays()) {
            val intent = overlayPermission().settingsIntent(application)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
            application.startActivity(intent)
        }
    }

    /** Temporary 2.x bridge retained until T05/T10 migrate geometry consumers. */
    @Deprecated("Obtain bounds and insets from the target window instead.")
    public fun getDisplayWidth(): Int = android.util.DisplayMetrics().let { metrics -> windowManager.defaultDisplay.getMetrics(metrics); metrics.widthPixels }

    /** Temporary 2.x bridge retained until T05/T10 migrate geometry consumers. */
    @Deprecated("Obtain bounds and insets from the target window instead.")
    public fun getDisplayHeight(): Int = android.util.DisplayMetrics().let { metrics -> windowManager.defaultDisplay.getMetrics(metrics); metrics.heightPixels }

    private fun <T : View> newApplicationOverlay(view: T, spec: OverlaySpec): OverlayView<T> {
        requireMainThread()
        require(view.parent == null) { "The managed view must not already have a parent." }
        require(spec.screenBrightness == null) { "screenBrightness is supported only for activity overlays." }
        return OverlayView(view, OverlayScope.APPLICATION, OverlayWindowManager.getApplicationInstance(), spec, permission)
    }

    private fun <T : View> newActivityOverlay(view: T, activity: Activity, spec: OverlaySpec): OverlayView<T> {
        requireMainThread()
        require(view.parent == null) { "The managed view must not already have a parent." }
        require(!activity.isFinishing && (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR1 || !activity.isDestroyed)) { "Activity is finishing or destroyed." }
        lateinit var handle: OverlayView<T>
        handle = OverlayView(
            view,
            OverlayScope.ACTIVITY,
            OverlayWindowManager.getActivityInstance(activity),
            spec,
            permission,
        ) { ActivityOverlayRegistry.deregister(activity, handle) }
        ActivityOverlayRegistry.register(activity, handle)
        return handle
    }

    private fun requireMainThread() {
        val main = Looper.getMainLooper()
        check(main != null && Looper.myLooper() === main) { "OverlayViewManager must be used on the main thread." }
    }

    public companion object {
        @Volatile private var instance: OverlayViewManager? = null

        /**
         * Registered on the first successful [init], never again on a same-Application no-op init.
         * Forces every live activity-scoped handle of the destroyed Activity to release and drops
         * the Activity's entry from [OverlayWindowManager]'s per-Activity instance cache.
         */
        private val activityDestructionCallbacks: Application.ActivityLifecycleCallbacks =
            object : SimpleActivityLifecycleCallbacks() {
                override fun onActivityDestroyed(activity: Activity) {
                    ActivityOverlayRegistry.disposeAll(activity)
                    OverlayWindowManager.removeActivityInstance(activity)
                }
            }

        /** Initializes the singleton. Repeating this with the same Application is harmless. */
        @JvmStatic @MainThread public fun init(application: Application) {
            val main = Looper.getMainLooper()
            check(main != null && Looper.myLooper() === main) { "OverlayViewManager must be initialized on the main thread." }
            synchronized(this) {
                val existing = instance
                check(existing == null || existing.application === application) { "OverlayViewManager is already initialized for another Application." }
                if (existing == null) {
                    instance = OverlayViewManager(application)
                    application.registerActivityLifecycleCallbacks(activityDestructionCallbacks)
                }
            }
        }

        /** Returns the initialized process singleton. */
        @JvmStatic @MainThread public fun getInstance(): OverlayViewManager {
            val main = Looper.getMainLooper()
            check(main != null && Looper.myLooper() === main) { "OverlayViewManager must be accessed on the main thread." }
            return instance ?: throw IllegalStateException("OverlayViewManager is not initialized. Call init(Application) first.")
        }

    }
}
